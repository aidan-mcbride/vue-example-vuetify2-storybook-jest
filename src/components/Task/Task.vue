<template>
  <v-list-item>
    <v-list-item-action>
      <v-btn
        class="task__archive-button"
        aria-label="archive task"
        text
        icon
        @click="$emit('archiveTask', task.id)"
      >
        <v-icon>{{ checkboxIcon }}</v-icon>
      </v-btn>
    </v-list-item-action>
    <v-list-item-content>
      <v-list-item-title>{{ task.title }}</v-list-item-title>
    </v-list-item-content>
    <v-list-item-action>
      <v-btn
        v-if="!isChecked"
        class="task__pin-button"
        aria-label="pin task"
        text
        icon
        @click="$emit('pinTask', task.id)"
      >
        <v-icon v-if="isPinned" color="blue">mdi-star</v-icon>
        <v-icon v-else>mdi-star-outline</v-icon>
      </v-btn>
    </v-list-item-action>
  </v-list-item>
</template>

<script>
export default {
  name: 'Task',
  props: {
    task: {
      type: Object,
      required: true,
      default: () => ({
        id: '',
        state: '',
        title: ''
      })
    }
  },
  computed: {
    isChecked() {
      return this.task.state === 'TASK_ARCHIVED';
    },
    isPinned() {
      return this.task.state === 'TASK_PINNED';
    },
    checkboxIcon() {
      return this.isChecked
        ? 'mdi-checkbox-marked-outline'
        : 'mdi-checkbox-blank-outline';
    }
  }
};
</script>
