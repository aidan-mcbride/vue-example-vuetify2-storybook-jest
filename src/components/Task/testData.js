export const taskData = {
  id: '1',
  title: 'Test Task',
  state: 'TASK_INBOX',
  updated_at: new Date(2019, 0, 1, 9, 0)
};
