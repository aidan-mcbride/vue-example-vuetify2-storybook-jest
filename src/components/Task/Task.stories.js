import { action } from '@storybook/addon-actions';
import Task from './Task';

import { taskData } from './testData';
export default {
  title: 'Task',
  // Our exports that end in "Data" are not stories.
  excludeStories: /.*Data$/
};
export const actionsData = {
  onPinTask: action('onPinTask'),
  onArchiveTask: action('onArchiveTask')
};

const taskTemplate = `<task :task="task" @archiveTask="onArchiveTask" @pinTask="onPinTask"/>`;

// default task state
export const Default = () => ({
  components: { Task },
  template: taskTemplate,
  props: {
    task: {
      default: () => taskData
    }
  },
  methods: actionsData
});
// pinned task state
export const Pinned = () => ({
  components: { Task },
  template: taskTemplate,
  props: {
    task: {
      default: () => ({
        ...taskData,
        state: 'TASK_PINNED'
      })
    }
  },
  methods: actionsData
});
// archived task state
export const Archived = () => ({
  components: { Task },
  template: taskTemplate,
  props: {
    task: {
      default: {
        ...taskData,
        state: 'TASK_ARCHIVED'
      }
    }
  },
  methods: actionsData
});
