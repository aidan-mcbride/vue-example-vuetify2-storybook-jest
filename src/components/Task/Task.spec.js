// Components
import Task from './Task';

// Utilities
import { mount } from '@vue/test-utils';

// Test Data
import { taskData } from './testData';

describe('Task', () => {
  describe('Default State', () => {
    it('renders correct markup', () => {
      const wrapper = mount(Task, {
        propsData: {
          task: taskData
        }
      });
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders bound task title', () => {
      // arrange
      const wrapper = mount(Task, {
        propsData: {
          task: taskData
        }
      });
      const title = wrapper.find('.v-list-item__title');

      // assert
      expect(title.text()).toBe(taskData.title);
    });

    it('emits archive event with task ID when archive button is clicked', () => {
      // arrange
      const wrapper = mount(Task, { propsData: { task: taskData } });
      const archiveButton = wrapper.find('.task__archive-button');

      // act
      archiveButton.trigger('click');

      // assert
      const actualValue = wrapper.emitted().archiveTask[0][0];
      const expectedValue = taskData.id;
      expect(actualValue).toBe(expectedValue);
    });

    it('emits pin task event with task ID when pin task button is clicked', () => {
      // arrange
      const wrapper = mount(Task, { propsData: { task: taskData } });
      const pinButton = wrapper.find('.task__pin-button');

      // act
      pinButton.trigger('click');

      // assert
      const actualValue = wrapper.emitted().pinTask[0][0];
      const expectedValue = taskData.id;
      expect(actualValue).toBe(expectedValue);
    });
  });

  describe('Pinned State', () => {
    it('renders correct markup', () => {
      const wrapper = mount(Task, {
        propsData: { task: { ...taskData, state: 'TASK_PINNED' } }
      });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('Archived State', () => {
    it('renders correct markup', () => {
      const wrapper = mount(Task, {
        propsData: { task: { ...taskData, state: 'TASK_ARCHIVED' } }
      });
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
