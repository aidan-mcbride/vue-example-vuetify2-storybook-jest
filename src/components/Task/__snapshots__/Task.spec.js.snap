// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Archived State renders correct markup 1`] = `
<div tabindex="-1" class="v-list-item theme--light">
  <div class="v-list-item__action"><button type="button" class="task__archive-button v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-marked-outline theme--light"></i></span></button></div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Test Task</div>
  </div>
  <div class="v-list-item__action">
    <!---->
  </div>
</div>
`;

exports[`Task Default State renders correct markup 1`] = `
<div tabindex="-1" class="v-list-item theme--light">
  <div class="v-list-item__action"><button type="button" class="task__archive-button v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"></i></span></button></div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Test Task</div>
  </div>
  <div class="v-list-item__action"><button type="button" class="task__pin-button v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star-outline theme--light"></i></span></button></div>
</div>
`;

exports[`Task Pinned State renders correct markup 1`] = `
<div tabindex="-1" class="v-list-item theme--light">
  <div class="v-list-item__action"><button type="button" class="task__archive-button v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-checkbox-blank-outline theme--light"></i></span></button></div>
  <div class="v-list-item__content">
    <div class="v-list-item__title">Test Task</div>
  </div>
  <div class="v-list-item__action"><button type="button" class="task__pin-button v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-star theme--light blue--text"></i></span></button></div>
</div>
`;
